File: backend/schemas/tax.py
from datetime import datetime

from pydantic import BaseModel


class BandBreakdownItem(BaseModel):
    rate: float
    amount_in_band: float
    tax: float


class CategoryAmountItem(BaseModel):
    category_name: str
    amount: float
    # Only populated for deduction_items and capital_allowance_items — the
    # pre-deduction amount and the effective rate applied to reach `amount`
    # (e.g. 100% for a normal expense, a user's home-office percentage for
    # utilities, or a First Schedule capital allowance class rate).
    gross_amount: float | None = None
    rate: float | None = None


class TaxComputationOut(BaseModel):
    tax_year: str
    total_income: float
    total_deductions: float
    total_reliefs: float
    total_capital_allowances: float
    taxable_income: float
    estimated_tax_owed: float
    minimum_wage_exempt: bool
    band_breakdown: list[BandBreakdownItem]
    income_items: list[CategoryAmountItem] = []
    deduction_items: list[CategoryAmountItem] = []
    relief_items: list[CategoryAmountItem] = []
    capital_allowance_items: list[CategoryAmountItem] = []
    last_updated: datetime | None

File: frontend/src/api/tax.ts
import { apiFetch } from '../lib/apiClient'

export interface BandBreakdownItem {
  rate: number
  amount_in_band: number
  tax: number
}

export interface CategoryAmountItem {
  category_name: string
  amount: number
}

export interface TaxComputation {
  tax_year: string
  total_income: number
  total_deductions: number
  total_reliefs: number
  total_capital_allowances: number
  taxable_income: number
  estimated_tax_owed: number
  minimum_wage_exempt: boolean
  band_breakdown: BandBreakdownItem[]
  income_items: CategoryAmountItem[]
  deduction_items: CategoryAmountItem[]
  relief_items: CategoryAmountItem[]
  capital_allowance_items: CategoryAmountItem[]
  last_updated: string | null
}

export function getTaxComputation(taxYear: string) {
  return apiFetch<TaxComputation>(`/tax-computations/${taxYear}`)
}

export function computeTax(taxYear: string) {
  return apiFetch<TaxComputation>(`/tax-computations/${taxYear}/compute`, { method: 'POST' })
}

export async function downloadReport(taxYear: string): Promise<Blob> {
  return apiFetch<Blob>(`/tax-computations/${taxYear}/report`)
}

export interface ActionItem {
  message: string
  href: string
}

export interface DashboardData {
  tax_year: string
  total_income: number
  total_deductions: number
  total_reliefs: number
  total_capital_allowances: number
  estimated_tax_owed: number
  pending_review_count: number
  locked_statements_count: number
  missing_receipts_count: number
  google_drive_connected: boolean
  filing_guidance: string
  outstanding_actions: ActionItem[]
}

export function getDashboard(taxYear: string) {
  return apiFetch<DashboardData>('/dashboard', { query: { tax_year: taxYear } })
}
